zubbi/scraper/scraper.py

```diff
@@ -146,39 +160,76 @@ def get_file_info(self, path):
 
     def scrape_role_files(self):
         role_files = {}
+        # Roles might be grouped in some parent directory, so we need to
+        # recursively check nested directories till we find role directory.
         # We are only interested in the role name (=> directory name)
         # and the README and CHANGELOG files. Thus, we don't need to
         # iterate recursively over the roles directory as those files
         # should be on the top-level per role.
+
+        # Listing repo files is based on the example in pyGithub doc:
+        # https://pygithub.readthedocs.io/en/v2.1.1/examples/Repository.html#get-all-of-the-contents-of-the-repository-recursively
         try:
-            roles = self.repo.directory_contents(ROLES_DIRECTORY)
-            for role_name, remote_file in roles.items():
+            # Ansible requires the role to be defined in a certain directory structure.
+            # Thus, we are only interested in directories. Files can be omitted
+            dir_contents = [
+                item
+                for item in self.repo.directory_contents(ROLES_DIRECTORY).values()
+                if item.type == "dir"
+            ]
+            while dir_contents:
```

Comment:
One drawback of the new solution seems to be that we now check each role twice: Once during the "initial" loop where we check out the contents of the `roles` directory (e.g. `roles/foo`) and a second time when we check if the directory actually contains a role ( e.g. `roles/foo/defaults`). Could we somehow combine this in a single step? I mean we don' actually need to check the contents of the `roles/foo/defaults` directory. It would be sufficient to check the contents of the `roles/foo` directory to see that it contains a `defaults` directory.

I think the problem is line [190](https://github.com/bmwcarit/zubbi/pull/108/files#diff-1410fee0224b9ce6de9250a205ce658a9162443f5f5f875e447987dced2e0ddbR190): In case of a role directory we are adding all subdirectories (`roles/foo/defaults`, `roles/foo/tasks`, `roles/foo/vars`, ...) to the `dir_contents` list to check out those directories later on. However, in this case we already have the information that this directory contains a role. So, instead of adding all subdirectories for this role to the search list, we could directly "parse" this role (check out the changelog/readme files, get the last_changed timestamp) and remove the role directory itself from the search list (this step might already be done in line [182](https://github.com/bmwcarit/zubbi/pull/108/files#diff-1410fee0224b9ce6de9250a205ce658a9162443f5f5f875e447987dced2e0ddbR182)).

Let me know if something is unclear.

Reply:
Thanks, this looks much simpler now.